lib/phoenix_live_view/tag_engine.ex

```diff
@@ -743,39 +773,186 @@ defmodule Phoenix.LiveView.TagEngine do
+            state
+            |> set_root_on_tag()
+            |> push_tag(token)
+            |> handle_tag_and_attrs(name, attrs, ">", to_location(tag_meta))
+            |> continue(tokens)
+
+          {true, new_meta, new_attrs} ->
+            state
+            |> push_substate_to_stack()
+            |> update_subengine(:handle_begin, [])
+            |> set_root_on_not_tag()
+            |> push_tag({:tag, name, new_attrs, new_meta})
+            |> handle_tag_and_attrs(name, new_attrs, ">", to_location(new_meta))
+            |> continue(tokens)
+        end
     end
   end
 
-  defp handle_token({:close, :tag, name, tag_meta} = token, state) do
+  defp handle_token([{:close, :tag, name, tag_meta} = token | tokens], state) do
     {{:tag, _name, _attrs, tag_open_meta}, state} = pop_tag!(state, token)
 
     state
     |> update_subengine(:handle_text, [to_location(tag_meta), "</#{name}>"])
     |> handle_special_expr(tag_open_meta)
+    |> continue(tokens)
+  end
+
+  defp handle_token([], state), do: state
+
+  defp handle_macro_component(
+         [{:tag, _name, _attrs, tag_meta} | _] = tokens,
+         module_string,
+         state
+       ) do
+    # Macro components work by converting the HEEx tokens into an AST
+    # (see Phoenix.Component.MacroComponent) and then calling the transform
+    # function on the macro component module, which can return a transformed
+    # AST.
+    #
+    # The AST is limited in functionality and we handle it separately in
+    # the handle_ast function.
+
+    Macro.Env.required?(state.caller, Phoenix.Component) ||
+      raise ArgumentError,
+            "macro components are only supported in modules that `use Phoenix.Component`"
+
+    module = validate_module!(module_string, tag_meta, state)
+
+    {ast, rest} =
+      case Phoenix.Component.MacroComponent.build_ast(tokens, state.caller) do
+        {:ok, ast, rest} -> {ast, rest}
+        {:error, message, meta} -> raise_syntax_error!(message, meta, state)
+      end
+
+    try do
+      module.transform(ast, %{env: state.caller})
+    rescue
+      e in ArgumentError ->
+        raise_syntax_error!(
+          Exception.message(e),
+          tag_meta,
+          state
+        )
+    else
+      {:ok, new_ast} ->
+        state
+        |> handle_ast(new_ast, tag_meta)
+        |> continue(rest)
+
+      {:ok, new_ast, data} ->
+        Module.put_attribute(state.caller.module, :__macro_components__, {module, data})
+
+        state
+        |> handle_ast(new_ast, tag_meta)
+        |> continue(rest)
+
+      other ->
+        raise ArgumentError,
+              "a macro component must return {:ok, ast} or {:ok, ast, data}, got: #{inspect(other)}"
+    end
+  end
+
+  # self closing / void tags cannot have children
+  defp handle_ast(state, {tag, attrs, [], %{closing: closing}}, tag_open_meta) do
+    suffix =
+      case closing do
+        :void -> ">"
+        :self -> "/>"
+      end
+
+    state
+    |> set_root_on_tag()
```

Comment:
You can skip the set_root_on_tag here and below I think, because we are not accumulating tags given the AST guarantees all tags are well-formed.

Reply:
that would mean you cannot use a macro component at the root of a live component, right? But I guess that doesn't seem like a frequent use case 🤷🏻‍♂️